File: Convert Sorted List to Binary Search Tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *sortedListToBST(ListNode *head) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        
        if(head==NULL)
        {
            return NULL;
        }
        
        ListNode *twosteps = head->next;
        ListNode *onestep = head;
        ListNode *pre = NULL;
        while(twosteps)
        {
            if(twosteps->next)
            {
                twosteps = twosteps->next->next;
            }
            else
            {
                twosteps = NULL;
            }
            pre = onestep;
            onestep = onestep->next;
        }
        
        TreeNode *root = new TreeNode(onestep->val);
        
        if(pre)
            pre->next = NULL;
        else
            head = pre;
            
        root->left = sortedListToBST(head);
        root->right = sortedListToBST(onestep->next);
        onestep->next = NULL;
        free(onestep);
        
        return root;
    }
    
  
};

File: Convert Sorted List to Binary Search Tree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; next = null; }
 * }
 */
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    ListNode curNode;
    public TreeNode sortedListToBST(ListNode head) {
        int num = 0;
        curNode = head;
        while(curNode != null)
        {
            num++;
            curNode = curNode.next;
        }
        curNode = head;
        return sortedListToBST(0, num-1);
    }
    
    public TreeNode sortedListToBST(int start, int end)
    {
        if(start> end)
        {
            return null;
        }
        
        int mid = start + (end - start)/2;
        TreeNode left = sortedListToBST(start, mid - 1);
        TreeNode root = new TreeNode(curNode.val);
        curNode = curNode.next;
        TreeNode right = sortedListToBST(mid + 1, end);
        
        root.left = left; 
        root.right = right;
        return root;
    }
}